plugins/lvmprovisioner/provisioner.go

```diff
@@ -0,0 +1,133 @@
+package lvmprovisioner
+
+import (
+	"errors"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/gluster/glusterd2/pkg/fsutils"
+	"github.com/gluster/glusterd2/pkg/lvmutils"
+)
+
+var mountOpts = "rw,inode64,noatime,nouuid"
+
+// Provisioner represents lvm provisioner plugin
+type Provisioner struct{}
+
+func getVgName(devpath string) string {
+	return "gluster" + strings.Replace(devpath, "/", "-", -1)
+}
+
+func getLvName(brickid string) string {
+	return "lv-" + brickid
+}
+
+func getTpName(brickid string) string {
+	return "tp-" + brickid
+}
+
+func getBrickDev(devpath, brickid string) string {
+	return "/dev/" + getVgName(devpath) + "/" + getLvName(brickid)
+}
+
+// Register creates pv and vg for a given device
+func (p Provisioner) Register(devpath string) error {
+	err := lvmutils.CreatePV(devpath)
+	if err != nil {
+		return err
+	}
+	return lvmutils.CreateVG(devpath, getVgName(devpath))
+}
+
+// AvailableSize returns available size in the given device
+func (p Provisioner) AvailableSize(devpath string) (uint64, uint64, error) {
+	return lvmutils.GetVgAvailableSize(getVgName(devpath))
+}
+
+// Unregister removes VG and PV
+func (p Provisioner) Unregister(devpath string) error {
+	err := lvmutils.RemoveVG(getVgName(devpath))
+	if err != nil {
+		return err
+	}
+	return lvmutils.RemovePV(devpath)
+}
+
+// CreateBrick creates thinpool and lv for given size
+func (p Provisioner) CreateBrick(devpath, brickid string, size uint64, bufferFactor float64) error {
+	vgname := getVgName(devpath)
+	tpsize := uint64(float64(size) * bufferFactor)
+	tpname := getTpName(brickid)
+	lvname := getLvName(brickid)
+	metasize := lvmutils.GetTpMetadataSize(tpsize)
+
+	err := lvmutils.CreateTP(vgname, tpname, tpsize, metasize)
+	if err != nil {
+		return err
+	}
+	return lvmutils.CreateLV(vgname, tpname, lvname, size)
+}
+
+// CreateBrickFS creates the filesystem as requested
+func (p Provisioner) CreateBrickFS(devpath, brickid, fstype string) error {
+	brickdev := getBrickDev(devpath, brickid)
+	switch fstype {
+	case "xfs":
+		return fsutils.MakeXfs(brickdev)
+	default:
+		return errors.New("unsupported filesystem")
+	}
+}
+
+// CreateBrickDir creates brick directory inside mount
+func (p Provisioner) CreateBrickDir(brickPath string) error {
+	return os.MkdirAll(brickPath, os.ModeDir|os.ModePerm)
+}
+
+// MountBrick mounts the brick to the given brick path
+func (p Provisioner) MountBrick(devpath, brickid, brickPath string) error {
+	mountdir := path.Dir(brickPath)
+	brickdev := getBrickDev(devpath, brickid)
+	err := os.MkdirAll(mountdir, os.ModeDir|os.ModePerm)
+	if err != nil {
+		return err
+	}
+	return fsutils.Mount(brickdev, mountdir, mountOpts)
```

Comment:
I think mount can be variable. Different provisioners of the same type may need to mount with different mount options. For example, in the case of a snapshot, we have to mount using the same options that the parent bricks are used. In this case, the default value won't help

Reply:
Snapshot mount should be different interface function in my opinion. Bricks of Cloned volume will use this function but not snapshot bricks.

Reply:
Any particular reason why snapshot volume should have a separate mount interface. There is nothing special for snapshot mount. IMHO, we can just take brickinfo.MountInfo as an argument . Becuase this strcture is contains information exclusive to mount,
eg :glusterd2/volume/fs_utils.go: MountDirectory

Reply:
Any particular reason why snapshot should have separate mount interface. Snapshot mount has nothing special comared to the normal mount. IMHO, we can take brickinfo.MountInfo as an argument, because it is a strcture contains xclusive information about mount. And that is common to any provisoner. Something like glusterd2/volume/fs_utils.go:MountDirectory